tests/test_permissions.py

```diff
@@ -0,0 +1,319 @@
+            "stage_moderator",
+            "events",
+            "advanced",
+            "private_channel",
+        ],
+    )
+    def test_classmethods(self, method_name: str):
+        method = getattr(Permissions, method_name)
+
+        perms: Permissions = method()
+        assert isinstance(perms, Permissions)
+
+        # check that caching does not return the same permissions instance
+        perms_two: Permissions = method()
+        assert perms is not perms_two
+        assert perms.value == perms_two.value
+
+
+class TestPermissionOverwrite:
+    def test_init(self) -> None:
+        perms = PermissionOverwrite(manage_messages=True)
+
+        assert perms.manage_messages is True
+
+    def test_init_invalid_perms(self) -> None:
+        with pytest.raises(ValueError, match="'h' is not a valid permission name."):
+            PermissionOverwrite(h=True)
+
+    def test_equality(self) -> None:
+        one = PermissionOverwrite()
+        two = PermissionOverwrite()
+
+        assert one is not two
+        assert one == two
+
+        two.ban_members = False
+        assert one != two
+
+    def test_set(self) -> None:
+        po = PermissionOverwrite()
+        po.attach_files = False
+        assert po.attach_files is False
+
+        po.attach_files = True
+        assert po.attach_files is True
+
+        po.attach_files = None
+        assert po.attach_files is None
+
+    def test_set_invalid_type(self) -> None:
+        po = PermissionOverwrite()
+        with pytest.raises(TypeError, match="Expected bool or NoneType, received str"):
+            po.connect = "h"  # type: ignore
+
+        with pytest.raises(
+            AttributeError, match="'PermissionOverwrite' object has no attribute 'oh'"
+        ):
+            po.oh = False  # type: ignore
+
+    @pytest.mark.parametrize(
+        ("allow", "deny"),
+        [
+            ({"view_channel": True}, {"ban_members": True}),
+            ({"view_channel": True}, {"view_channel": True}),
+            ({"administrator": True}, {"manage_channels": False}),
+        ],
+    )
+    def test_from_pair(
+        self,
+        allow: Dict[str, bool],
+        deny: Dict[str, bool],
+    ) -> None:
+        perm_allow = Permissions(**allow)
+        perm_deny = Permissions(**deny)
+
+        po = PermissionOverwrite.from_pair(perm_allow, perm_deny)
+
+        # iterate over the allowed perms and assert that the overwrite is what the allowed perms are
+        for perm, allowed in perm_allow:
+            # get the attr from the denied perms as denied perms override the allow list in from_pair
+            if allowed and not getattr(perm_deny, perm):
+                assert getattr(po, perm) is True
+            else:
+                assert getattr(po, perm) is not True
+
+        for perm, denied in deny.items():
+            if denied:
+                assert getattr(po, perm) is False
+            else:
+                assert getattr(po, perm) is not False
```

Comment:
I *think* we can simplify that test to this? Correct me if I'm wrong though.
```suggestion
    @pytest.mark.parametrize(
        ("allow", "deny", "expected"),
        [
            ({"view_channel": True}, {"ban_members": True}, {"view_channel": True, "ban_members": False}),
            ({"view_channel": True}, {"view_channel": True}, {"view_channel": False}),
            ({"administrator": True}, {"manage_channels": False}, {"administrator": True}),
        ],
    )
    def test_from_pair(
        self,
        allow: Dict[str, bool],
        deny: Dict[str, bool],
        expected: Dict[str, bool],
    ) -> None:
        perm_allow = Permissions(**allow)
        perm_deny = Permissions(**deny)

        po = PermissionOverwrite.from_pair(perm_allow, perm_deny)

        assert {k: v for k, v in po if v is not None} == expected
```

Reply:
I'd rather leave it how it is because I think it is a bit more readable, but I'll adopt a few of the suggestions shown here.

Reply:
https://github.com/DisnakeDev/disnake/pull/609/commits/74364dfa78b3b1bc4c60229b12f15652f3d8a57e